### PR TITLE
mrc-2528 Complete ShinyRob prototype

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch fixes most of the issues remaining in the prototype at the end of the last sprint. I'd suggest that the fact that this now basically works as required means that we can indeed treat the ShinyRob chart as a 'generic chart' as implemented here, rather than needing to create a custom component with special knowledge of this particular chart. The only 'special' added here is to deal with variable numbers of subplots, indicated by optional chart metadata. 

Changes made:
- fixes bug where first two subplots shared axes (need to get axis names rights, and also don't show more plots than you have defined in rows and columns settings)
- Colours line segments exceeding threshold value drop or rise in red, all other segments are black. (Not implemnted for barchart, as barchart does not need to be included in first version.) We are not including user editing of threshold for first version either - this is hard-coded to 0.25 increase or decrease.
- Implements scroll of subplots area (this seemed to be preferred to paging by the Naomi group), and no longer limits number of plots to 30. A special 'subplots' config element can now  be included, and if so the `ChartSlot`component uses it to calculate the number of subplot rows which will be required. This is passed to the Chart jsonata as additional metadata, along with the config-provided columns value, and the ChartSlot also uses it to calculate the required scroll height to contain all the rows. The ChartSlot also has a prop for `chartHeight` the height of the chart component on the page (container of the scrolled area). It's currently showing 3 plots per row as that seemed comfortable for this layout, but we can change that to 4 as in ShinyRob if preferred. 
- Area name and id are displayed as annotations on the plots, using the domains of subplot axes as reference points
- Adds some visual tweaks to the axes e.g. tick text in grey, and range goes to 0 for y axes
- Improved jsonata performance by using variables for values which were repeatedly calculated - load time was reduced from ~12s to 4.5s for me. In the real app, it would probably be worth showing a spinner while re-rendering.

This is not yet using the realistic sample data, but I think this should have minimal impact structurally - it will affect the filters somewhat e.g. will have basic area level filter rather than full Area filter (hence I haven't addressed the Area filter bugs). See ticket for discussion of other changes I'm proposing to leave until integration with HINT.

It occurs to me now that perhaps we should make `ChartSlot` (probably renamed!) a standalone component, incorporating the functionality of the `genericCharts` module, so that it does its own filtering etc. This feels easier now that we know the HINT datasets for all the new charts are likely to mean fetching non-standard datasets from a configured url (something we could include in the component) but it would also be possible if data was coming from the store e.g. by supporting specifying a standard data source by a name which maps to a Root getter.  

Making the component standalone in this way will make it easier to use in other applications including an application for quickly testing chart configs.

I haven't yet updated the KB article which references this prototype, but I will when this gets merged.

I haven't added all  changes into the bar plot as only the scatter plot is required for initial version.  

We'll probably remove most of the Plotly gubbins at the top of the chart (Zoom etc) but could leave e.g. download plot as image

### CODE REVIEW

File: src/components/ChartSlot.vue [hunk 1]
```diff
@@ -189,3 +217,25 @@
...
+    .chart-container {
+        overflow-y: scroll;
+        width: 100%;
+    }
+
+    .chart {
+        width: 100%;
+    }
+
+    .annotation-text {
```
Comment (EmmaLRussell) [inline src/components/ChartSlot.vue:230]: Plotly doesn't really provide a nice way to change font weight on annotations in config, so doing it through css here, though it should really be part of the chart configuration

**Review by mwoodbri — APPROVED** [review]
This is really impressive - I had my doubts over whether it would even be possible to implement this chart using Plotly - especially the gradients - but it looks great and I think the performance is acceptable given the complexity of the plot (+ spinner and possibly further optimisations).

I know you mentioned it, but I think we will need a short/updated summary of how the pieces fit together, and maybe a couple of bullet points explaining the most abstruse Jsonata syntax used (`&$` etc).

**Comment by EmmaLRussell** [comment]
> I know you mentioned it, but I think we will need a short/updated summary of how the pieces fit together, and maybe a couple of bullet points explaining the most abstruse Jsonata syntax used (`&$` etc).

Yes, I'll do that shortly, and great idea to also explain the Jsonata, not least as I'll definitely have forgotten how all that works before too long!

**Comment by Nicolas-Dolan** [comment]
This is very nice and seems very responsive. I gather from above that not showing axis labels and chart headers is intentional at this stage? I prefer to be able to read charts at a glance without having to interact with them individually to see what they are showing. Not something that needs to be worried about in this ticket though.

**Review by Nicolas-Dolan — APPROVED** [review]

**Review by LekanAnni — CHANGES_REQUESTED** [review]
Brilliant work !

The chart is looking great. However, the only thing I can pick up is its responsiveness, I think it should be responsive. If screen is reduced then users have to scroll right to see all columns, this should automatically display 1 column when screen becomes too small to display 3 chart columns.

**Comment by EmmaLRussell** [comment]
> This is very nice and seems very responsive. I gather from above that not showing axis labels and chart headers is intentional at this stage? I prefer to be able to read charts at a glance without having to interact with them individually to see what they are showing. Not something that needs to be worried about in this ticket though.

I think showing axis labels on every subplot would just take up too much real estate, user should understand meaning from context. Each subplot should now have its own header, showing area name and id - were you not seeing that?

**Comment by EmmaLRussell** [comment]
> The chart is looking great. However, the only thing I can pick up is its responsiveness, I think it should be responsive. If screen is reduced then users have to scroll right to see all columns, this should automatically display 1 column when screen becomes too small to display 3 chart columns.

That would indeed be nice - Plotly won't help us with that at all I don't think sadly. 
There's a couple of use cases here:
1. Responding to width changes by redrawing plots with configured number of columns to take up new available width
2. Configuring 'minimum chart width' and using that to override configured number of columns as you describe.

I'm actually not sure what the preferred use case here is - Rachel mentioned wanting to keep default area ordering so that the users had a fixed layout from ShinyRob and could always expect to see the same area in the same position - so it might actually be preferable to scroll as currently when space is limited. I will check!

**Comment by EmmaLRussell** [comment]
I've put in some basic responsiveness in the Plotly layout so that it at least responds to changes in width of the parent by resizing the plots (albeit slowly!) Unfortunately this means that at narrow widths the titles overflow - not clear how to fix this in Plotly. I've created a ticket to deal with this in the app when requirements are clarified: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2538
Would be good to get this branch merged if you're happy to leave the prototype at that.

**Review by LekanAnni — APPROVED** [review]

**Comment by LekanAnni** [comment]
> I've put in some basic responsiveness in the Plotly layout so that it at least responds to changes in width of the parent by resizing the plots (albeit slowly!) Unfortunately this means that at narrow widths the titles overflow - not clear how to fix this in Plotly. I've created a ticket to deal with this in the app when requirements are clarified: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2538
> Would be good to get this branch merged if you're happy to leave the prototype at that.



> I've put in some basic responsiveness in the Plotly layout so that it at least responds to changes in width of the parent by resizing the plots (albeit slowly!) Unfortunately this means that at narrow widths the titles overflow - not clear how to fix this in Plotly. I've created a ticket to deal with this in the app when requirements are clarified: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2538
> Would be good to get this branch merged if you're happy to leave the prototype at that.

Yeah I reckon mrc-2538 will address the responsiveness further. Thank you for making some changes.

**Pull request merged**